### ISSUE TITLE
Instance -> bytes -> instance conversion fails

### ISSUE THREAD

**Issue by sterliakov** [danielgtaylor/python-betterproto#414]
This was discovered when porting existing TypeScript library that uses protobuf (with [protoc-gen-ts](https://github.com/thesayyn/protoc-gen-ts/blob/main/src/field.ts), it it matters) heavily. 

I noticed minor binary incompatibilities in some messages, and also decode+encode composition is not an identity function (`Obj().parse(bytes(obj)) != obj`). 

The following (docstrings removed for consciousness) python code is generated from this source:
```
syntax = "proto3";

message TxRaw {
  bytes body_bytes = 1;
  bytes auth_info_bytes = 2;
  repeated bytes signatures = 3;
}
```

```python
from dataclasses import dataclass
from typing import List

import betterproto


@dataclass(eq=False, repr=False)
class TxRaw(betterproto.Message):
    body_bytes: bytes = betterproto.bytes_field(1)
    auth_info_bytes: bytes = betterproto.bytes_field(2)
    signatures: List[bytes] = betterproto.bytes_field(3)


body = bytes([
    10, 136, 1, 10, 28, 47, 99, 111, 115, 109, 111, 115, 46, 98, 97, 110, 107,
    46, 118, 49, 98, 101, 116, 97, 49, 46, 77, 115, 103, 83, 101, 110, 100,
    18, 104, 10, 43, 101, 116, 104, 109, 49, 116, 102, 101, 103, 102, 53, 48,
    110, 53, 120, 108, 48, 104, 100, 53, 99, 120, 102, 122, 106, 99, 97, 51,
    121, 108, 115, 102, 112, 103, 48, 102, 110, 101, 100, 53, 103, 113, 109,
    18, 43, 101, 116, 104, 109, 49, 116, 102, 101, 103, 102, 53, 48, 110, 53,
    120, 108, 48, 104, 100, 53, 99, 120, 102, 122, 106, 99, 97, 51, 121, 108,
    115, 102, 112, 103, 48, 102, 110, 101, 100, 53, 103, 113, 109, 26, 12, 10,
    7, 97, 112, 104, 111, 116, 111, 110, 18, 1, 49, 250, 63, 161, 1, 10, 42,
    47, 101, 116, 104, 101, 114, 109, 105, 110, 116, 46, 116, 121, 112, 101,
    115, 46, 118, 49, 46, 69, 120, 116, 101, 110, 115, 105, 111, 110, 79, 112,
    116, 105, 111, 110, 115, 87, 101, 98, 51, 84, 120, 18, 115, 8, 168, 70,
    18, 43, 101, 116, 104, 109, 49, 116, 102, 101, 103, 102, 53, 48, 110, 53,
    120, 108, 48, 104, 100, 53, 99, 120, 102, 122, 106, 99, 97, 51, 121, 108,
    115, 102, 112, 103, 48, 102, 110, 101, 100, 53, 103, 113, 109, 26, 65, 95,
    219, 33, 190, 98, 109, 191, 101, 139, 33, 77, 174, 170, 109, 101, 137,
    240, 235, 232, 240, 220, 190, 187, 173, 250, 236, 242, 67, 74, 82, 132,
    204, 12, 178, 164, 122, 106, 195, 154, 165, 178, 255, 188, 143, 234, 77,
    242, 13, 51, 173, 224, 246, 187, 119, 101, 170, 32, 217, 208, 71, 26, 172,
    211, 142, 27,
])
auth = bytes([
    10, 89, 10, 79, 10, 40, 47, 101, 116, 104, 101, 114, 109, 105, 110, 116,
    46, 99, 114, 121, 112, 116, 111, 46, 118, 49, 46, 101, 116, 104, 115, 101,
    99, 112, 50, 53, 54, 107, 49, 46, 80, 117, 98, 75, 101, 121, 18, 35, 10,
    33, 2, 4, 240, 251, 139, 244, 117, 162, 43, 198, 195, 82, 91, 129, 92, 67,
    226, 40, 138, 120, 15, 177, 225, 112, 28, 237, 67, 158, 203, 50, 160, 234,
    153, 18, 4, 10, 2, 8, 127, 24, 1, 18, 19, 10, 13, 10, 7, 97, 112, 104,
    111, 116, 111, 110, 18, 2, 50, 48, 16, 192, 154, 12,
])
sig = [b'']


tx_raw_common = bytes([
    10, 176, 2, 10, 136, 1, 10, 28, 47, 99, 111, 115, 109, 111, 115, 46, 98,
    97, 110, 107, 46, 118, 49, 98, 101, 116, 97, 49, 46, 77, 115, 103, 83,
    101, 110, 100, 18, 104, 10, 43, 101, 116, 104, 109, 49, 116, 102, 101,
    103, 102, 53, 48, 110, 53, 120, 108, 48, 104, 100, 53, 99, 120, 102, 122,
    106, 99, 97, 51, 121, 108, 115, 102, 112, 103, 48, 102, 110, 101, 100, 53,
    103, 113, 109, 18, 43, 101, 116, 104, 109, 49, 116, 102, 101, 103, 102,
    53, 48, 110, 53, 120, 108, 48, 104, 100, 53, 99, 120, 102, 122, 106, 99,
    97, 51, 121, 108, 115, 102, 112, 103, 48, 102, 110, 101, 100, 53, 103,
    113, 109, 26, 12, 10, 7, 97, 112, 104, 111, 116, 111, 110, 18, 1, 49, 250,
    63, 161, 1, 10, 42, 47, 101, 116, 104, 101, 114, 109, 105, 110, 116, 46,
    116, 121, 112, 101, 115, 46, 118, 49, 46, 69, 120, 116, 101, 110, 115,
    105, 111, 110, 79, 112, 116, 105, 111, 110, 115, 87, 101, 98, 51, 84, 120,
    18, 115, 8, 168, 70, 18, 43, 101, 116, 104, 109, 49, 116, 102, 101, 103,
    102, 53, 48, 110, 53, 120, 108, 48, 104, 100, 53, 99, 120, 102, 122, 106,
    99, 97, 51, 121, 108, 115, 102, 112, 103, 48, 102, 110, 101, 100, 53, 103,
    113, 109, 26, 65, 95, 219, 33, 190, 98, 109, 191, 101, 139, 33, 77, 174,
    170, 109, 101, 137, 240, 235, 232, 240, 220, 190, 187, 173, 250, 236, 242,
    67, 74, 82, 132, 204, 12, 178, 164, 122, 106, 195, 154, 165, 178, 255,
    188, 143, 234, 77, 242, 13, 51, 173, 224, 246, 187, 119, 101, 170, 32,
    217, 208, 71, 26, 172, 211, 142, 27, 18, 112, 10, 89, 10, 79, 10, 40, 47,
    101, 116, 104, 101, 114, 109, 105, 110, 116, 46, 99, 114, 121, 112, 116,
    111, 46, 118, 49, 46, 101, 116, 104, 115, 101, 99, 112, 50, 53, 54, 107,
    49, 46, 80, 117, 98, 75, 101, 121, 18, 35, 10, 33, 2, 4, 240, 251, 139,
    244, 117, 162, 43, 198, 195, 82, 91, 129, 92, 67, 226, 40, 138, 120, 15,
    177, 225, 112, 28, 237, 67, 158, 203, 50, 160, 234, 153, 18, 4, 10, 2, 8,
    127, 24, 1, 18, 19, 10, 13, 10, 7, 97, 112, 104, 111, 116, 111, 110, 18,
    2, 50, 48, 16, 192, 154, 12,
])
tx_raw_1 = bytes([
    *tx_raw_common,
    26,  # This differs
    0,
])
tx_raw_2 = bytes([
    *tx_raw_common,
    10,  # This differs
    0,
])


def test_tx_raw_bytes2obj_1():
    # OK
    obj = TxRaw().parse(tx_raw_1)
    assert obj.auth_info_bytes == auth
    assert obj.body_bytes == body
    assert obj.signatures == sig


def test_tx_raw_bytes2obj_2():
    # FAILS
    obj = TxRaw().parse(tx_raw_2)
    assert obj.auth_info_bytes == auth
    assert obj.body_bytes == body
    assert obj.signatures == sig


def test_tx_raw_obj2bytes_1():
    # FAILS
    obj = TxRaw(body, auth, sig)
    assert bytes(obj).hex() == tx_raw_1.hex()


def test_tx_raw_obj2bytes_2():
    # OK
    obj = TxRaw(body, auth, sig)
    assert bytes(obj).hex() == tx_raw_2.hex()
```

Output:
```
=========================================================================================================== test session starts ============================================================================================================
platform linux -- Python 3.10.2, pytest-7.1.2, pluggy-1.0.0 -- /home/stas/Documents/Work/pyevmos/venv/bin/python
cachedir: .pytest_cache
rootdir: /home/stas/Documents/Work/pyevmos, configfile: pyproject.toml
plugins: web3-5.30.0, cov-3.0.0
collected 22 items / 18 deselected / 4 selected                                                                                                                                                                                            

tests/test_example.py::test_tx_raw_bytes2obj_1 PASSED                                                                                                                                                                                [ 25%]
tests/test_example.py::test_tx_raw_bytes2obj_2 FAILED                                                                                                                                                                                [ 50%]
tests/test_example.py::test_tx_raw_obj2bytes_1 FAILED                                                                                                                                                                                [ 75%]
tests/test_example.py::test_tx_raw_obj2bytes_2 PASSED                                                                                                                                                                                [100%]

================================================================================================================= FAILURES =================================================================================================================
_________________________________________________________________________________________________________ test_tx_raw_bytes2obj_2 __________________________________________________________________________________________________________
tests/test_example.py:101: in test_tx_raw_bytes2obj_2
    assert obj.body_bytes == body
E   AssertionError: assert b'' == (b'\n\x88\x01\n\x1c/cosmos.bank.v1beta1.MsgSend\x12h\n+ethm1tfegf50n5xl0hd5cxf'\n b'zjca3ylsfpg0fned5gqm\x12+ethm1tfegf50n5xl0hd5cxfzjca3ylsfpg0fned5gq'\n b'm\x1a\x0c\n\x07aphoton\x12\x011\xfa?\xa1\x01\n*/ethermint.types.v1.Extensi'\n b'onOptionsWeb3Tx\x12s\x08\xa8F\x12+ethm1tfegf50n5xl0hd5cxfzjca3ylsfpg0fned5gq'\n b'm\x1aA_\xdb!\xbebm\xbfe\x8b!M\xae\xaame\x89\xf0\xeb\xe8\xf0\xdc'\n b'\xbe\xbb\xad\xfa\xec\xf2CJR\x84\xcc\x0c\xb2\xa4zj\xc3\x9a\xa5\xb2'\n b'\xff\xbc\x8f\xeaM\xf2\r3\xad\xe0\xf6\xbbwe\xaa \xd9\xd0G\x1a\xac\xd3\x8e\x1b')                                                                                                                                         
E     Full diff:
E     + b''
E     - (
E     -  b'\n\x88\x01\n\x1c/cosmos.bank.v1beta1.MsgSend\x12h\n+ethm1tfegf50n5xl0hd5cxf'
E     -  b'zjca3ylsfpg0fned5gqm\x12+ethm1tfegf50n5xl0hd5cxfzjca3ylsfpg0fned5gq'
E     -  b'm\x1a\x0c\n\x07aphoton\x12\x011\xfa?\xa1\x01\n*/ethermint.types.v1.Extensi'
E     -  b'onOptionsWeb3Tx\x12s\x08\xa8F\x12+ethm1tfegf50n5xl0hd5cxfzjca3ylsfpg0fned5gq'
E     -  b'm\x1aA_\xdb!\xbebm\xbfe\x8b!M\xae\xaame\x89\xf0\xeb\xe8\xf0\xdc'
E     -  b'\xbe\xbb\xad\xfa\xec\xf2CJR\x84\xcc\x0c\xb2\xa4zj\xc3\x9a\xa5\xb2'
E     -  b'\xff\xbc\x8f\xeaM\xf2\r3\xad\xe0\xf6\xbbwe\xaa \xd9\xd0G\x1a\xac\xd3\x8e\x1b',
E     - )
_________________________________________________________________________________________________________ test_tx_raw_obj2bytes_1 __________________________________________________________________________________________________________
tests/test_example.py:108: in test_tx_raw_obj2bytes_1
    assert bytes(obj).hex() == tx_raw_1.hex()
E   AssertionError: assert '0ab0020a88010a1c2f636f736d6f732e62616e6b2e763162657461312e4d736753656e6412680a2b6574686d31746665676635306e35786c306864356378667a6a636133796c7366706730666e65643567716d122b6574686d31746665676635306e35786c306864356378667a6a636133796c7366706730666e65643567716d1a0c0a076170686f746f6e120131fa3fa1010a2a2f65746865726d696e742e74797065732e76312e457874656e73696f6e4f7074696f6e73576562335478127308a846122b6574686d31746665676635306e35786c306864356378667a6a636133796c7366706730666e65643567716d1a415fdb21be626dbf658b214daeaa6d6589f0ebe8f0dcbebbadfaecf2434a5284cc0cb2a47a6ac39aa5b2ffbc8fea4df20d33ade0f6bb7765aa20d9d0471aacd38e1b12700a590a4f0a282f65746865726d696e742e63727970746f2e76312e657468736563703235366b312e5075624b657912230a210204f0fb8bf475a22bc6c3525b815c43e2288a780fb1e1701ced439ecb32a0ea9912040a02087f180112130a0d0a076170686f746f6e1202323010c09a0c0a00' == '0ab0020a88010a1c2f636f736d6f732e62616e6b2e763162657461312e4d736753656e6412680a2b6574686d31746665676635306e35786c306864356378667a6a636133796c7366706730666e65643567716d122b6574686d31746665676635306e35786c306864356378667a6a636133796c7366706730666e65643567716d1a0c0a076170686f746f6e120131fa3fa1010a2a2f65746865726d696e742e74797065732e76312e457874656e73696f6e4f7074696f6e73576562335478127308a846122b6574686d31746665676635306e35786c306864356378667a6a636133796c7366706730666e65643567716d1a415fdb21be626dbf658b214daeaa6d6589f0ebe8f0dcbebbadfaecf2434a5284cc0cb2a47a6ac39aa5b2ffbc8fea4df20d33ade0f6bb7765aa20d9d0471aacd38e1b12700a590a4f0a282f65746865726d696e742e63727970746f2e76312e657468736563703235366b312e5075624b657912230a210204f0fb8bf475a22bc6c3525b815c43e2288a780fb1e1701ced439ecb32a0ea9912040a02087f180112130a0d0a076170686f746f6e1202323010c09a0c1a00'                                                                                                                                                                 
E     - 0ab0020a88010a1c2f636f736d6f732e62616e6b2e763162657461312e4d736753656e6412680a2b6574686d31746665676635306e35786c306864356378667a6a636133796c7366706730666e65643567716d122b6574686d31746665676635306e35786c306864356378667a6a636133796c7366706730666e65643567716d1a0c0a076170686f746f6e120131fa3fa1010a2a2f65746865726d696e742e74797065732e76312e457874656e73696f6e4f7074696f6e73576562335478127308a846122b6574686d31746665676635306e35786c306864356378667a6a636133796c7366706730666e65643567716d1a415fdb21be626dbf658b214daeaa6d6589f0ebe8f0dcbebbadfaecf2434a5284cc0cb2a47a6ac39aa5b2ffbc8fea4df20d33ade0f6bb7765aa20d9d0471aacd38e1b12700a590a4f0a282f65746865726d696e742e63727970746f2e76312e657468736563703235366b312e5075624b657912230a210204f0fb8bf475a22bc6c3525b815c43e2288a780fb1e1701ced439ecb32a0ea9912040a02087f180112130a0d0a076170686f746f6e1202323010c09a0c1a00                                                                                          
E     ?                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           ^^^^                                                                                          
E     + 0ab0020a88010a1c2f636f736d6f732e62616e6b2e763162657461312e4d736753656e6412680a2b6574686d31746665676635306e35786c306864356378667a6a636133796c7366706730666e65643567716d122b6574686d31746665676635306e35786c306864356378667a6a636133796c7366706730666e65643567716d1a0c0a076170686f746f6e120131fa3fa1010a2a2f65746865726d696e742e74797065732e76312e457874656e73696f6e4f7074696f6e73576562335478127308a846122b6574686d31746665676635306e35786c306864356378667a6a636133796c7366706730666e65643567716d1a415fdb21be626dbf658b214daeaa6d6589f0ebe8f0dcbebbadfaecf2434a5284cc0cb2a47a6ac39aa5b2ffbc8fea4df20d33ade0f6bb7765aa20d9d0471aacd38e1b12700a590a4f0a282f65746865726d696e742e63727970746f2e76312e657468736563703235366b312e5075624b657912230a210204f0fb8bf475a22bc6c3525b815c43e2288a780fb1e1701ced439ecb32a0ea9912040a02087f180112130a0d0a076170686f746f6e1202323010c09a0c0a00                                                                                          
E     ?                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                           ^^^^                                                                                          
========================================================================================================= short test summary info ==========================================================================================================
FAILED tests/test_example.py::test_tx_raw_bytes2obj_2 - AssertionError: assert b'' == (b'\n\x88\x01\n\x1c/cosmos.bank.v1beta1.MsgSend\x12h\n+ethm1tfegf50n5xl0hd5cxf'\n b'zjca3ylsfpg0fned5gqm\x12+ethm1tfegf50n5xl0hd5cxfzjca3ylsfpg0fne...
FAILED tests/test_example.py::test_tx_raw_obj2bytes_1 - AssertionError: assert '0ab0020a88010a1c2f636f736d6f732e62616e6b2e763162657461312e4d736753656e6412680a2b6574686d31746665676635306e35786c306864356378667a6a636133796c7366706730666...
================================================================================================ 2 failed, 2 passed, 18 deselected in 2.20s ================================================================================================
```

As you may see, tx_raw_1 can be successfully deserialized, but cannot be obtained via serialization. tx_raw_2 cannot be deserialized, but is created via serialization, so instance -> bytes -> instance conversion does not result in original instance (`instance.parse()` is not an inverse of `bytes(instance)`. `tx_raw_1` is the value borrowed from original test suite in TypeScript, so I suppose it is the proper one.

Could you drop any clue on what I'm missing?

I have also an example of another wrong (?) serialization, which is not 100% related to this.

Versions affected: at least 2.0.0b5, 2.0.0b4 and 1.2.5, I haven't tried any other.
`protoc 3.21.3; python 3.10 (from source); Ubuntu 22.04`

**Comment by sterliakov** [comment]
If I change `_serialize_single` definition default to use `serialize_empty = True` or pass it in `Message.__bytes__` (in `if isinstance(value, list) > else` branch), this works as expected, so something weird is happening. During serialization empty fields are not serialized, but these fields are expected during deserialization. Or is it intended behavior and these functions should not give identity in composition?

**Comment by Gobot1234** [comment]
Hi, thanks for the bug report but, I'm a little lost as to what the issue here is (would it be possible to use a smaller byte string?), can google's protobuf implementation parse this?

<img width="1197" alt="image" src="https://user-images.githubusercontent.com/50501825/187560914-56fba462-7a23-484e-8b33-e2e89a32577c.png">
(1 is on the top and 2 on the bottom)

It appears that 1 is actually parsable (with betterproto as well as with this app) and 2 isn't.

The round tripping of betterproto is tested I thought pretty thoroughly (https://github.com/danielgtaylor/python-betterproto/blob/master/tests/test_features.py) but, if this is not the case can you create an example that fails as it seems like the input you are giving betterproto is just bad.

**Comment by sterliakov** [comment]
1st is parseable by google protobuf and betterproto, 2nd is not, you're right. The problem is that betterproto actually generates 2nd byte string, as shown by tests in snippet, and then parses it incorrectly (expected: 1st is valid and 2nd is not). 

tx_raw_1 and tx_raw_2 are two byte strings that differ in one byte. To avoid manual comparison, I added 4 tests: one pair trying to decode both strings and compare to expected result, another pair encoding an object and comparing to byte strings. Tests show that `TxRaw().parse(tx_raw_1) == TxRaw(*args)` and `bytes(TxRaw(*args)) == tx_raw_2`, which is an unexpected (for me - I'm quite new to protobuf) result.

Also (this is a clue, IMO) betterproto does not match `protobuf` behavior here. When I run the same tests against code generated by protoc with `--python_out`, both encoding and decoding work properly for 1st bytes and fail for 2nd (this is what I expect).

Here's the code autogenerated by protobuf for python and corresponding tests (definitions same as before):

```python
# -*- coding: utf-8 -*-
# Generated by the protocol buffer compiler.  DO NOT EDIT!
# source: a.proto
"""Generated protocol buffer code."""
from google.protobuf.internal import builder as _builder
from google.protobuf import descriptor as _descriptor
from google.protobuf import descriptor_pool as _descriptor_pool
from google.protobuf import symbol_database as _symbol_database
# @@protoc_insertion_point(imports)

_sym_db = _symbol_database.Default()




DESCRIPTOR = _descriptor_pool.Default().AddSerializedFile(b'\n\x07\x61.proto\"H\n\x05TxRaw\x12\x12\n\nbody_bytes\x18\x01 \x01(\x0c\x12\x17\n\x0f\x61uth_info_bytes\x18\x02 \x01(\x0c\x12\x12\n\nsignatures\x18\x03 \x03(\x0c\x62\x06proto3')

_builder.BuildMessageAndEnumDescriptors(DESCRIPTOR, globals())
_builder.BuildTopDescriptorsAndMessages(DESCRIPTOR, 'a_pb2', globals())
if _descriptor._USE_C_DESCRIPTORS == False:

  DESCRIPTOR._options = None
  _TXRAW._serialized_start=11
  _TXRAW._serialized_end=83
# @@protoc_insertion_point(module_scope)


def test_tx_raw_bytes2obj_1():
    # OK
    obj = TxRaw.FromString(tx_raw_1)
    assert obj.auth_info_bytes == auth
    assert obj.body_bytes == body
    assert obj.signatures == sig


def test_tx_raw_bytes2obj_2():
    # FAILS
    obj = TxRaw.FromString(tx_raw_2)
    assert obj.auth_info_bytes == auth
    assert obj.body_bytes == body
    assert obj.signatures == sig


def test_tx_raw_obj2bytes_1():
    # OK (!)
    obj = TxRaw(body_bytes=body, auth_info_bytes=auth, signatures=sig)
    assert bytes(obj.SerializeToString()).hex() == tx_raw_1.hex()


def test_tx_raw_obj2bytes_2():
    # FAILS (!)
    obj = TxRaw(body_bytes=body, auth_info_bytes=auth, signatures=sig)
    assert bytes(obj.SerializeToString()).hex() == tx_raw_2.hex()
```
```
=========================================================================================================== test session starts ============================================================================================================
platform linux -- Python 3.10.2, pytest-7.1.2, pluggy-1.0.0
rootdir: /tmp/tests
collected 4 items                                                                                                                                                                                                                          

test_1.py .F.F                                                                                                                                                                                                                       [100%]

================================================================================================================= FAILURES =================================================================================================================
_________________________________________________________________________________________________________ test_tx_raw_bytes2obj_2 __________________________________________________________________________________________________________

    def test_tx_raw_bytes2obj_2():
        # FAILS
        obj = TxRaw.FromString(tx_raw_2)
        assert obj.auth_info_bytes == auth
>       assert obj.body_bytes == body
E       AssertionError: assert b'' == b'\n\x88\x01\...c\xd3\x8e\x1b'
E         Use -v to get more diff

test_1.py:116: AssertionError
_________________________________________________________________________________________________________ test_tx_raw_obj2bytes_2 __________________________________________________________________________________________________________

    def test_tx_raw_obj2bytes_2():
        # OK
        obj = TxRaw(body_bytes=body, auth_info_bytes=auth, signatures=sig)
>       assert bytes(obj.SerializeToString()).hex() == tx_raw_2.hex()
E       AssertionError: assert '0ab0020a8801...010c09a0c1a00' == '0ab0020a8801...010c09a0c0a00'
E         Skipping 832 identical leading characters in diff, use -v to show
E         - 3010c09a0c0a00
E         ?           ^
E         + 3010c09a0c1a00
E         ?           ^

test_1.py:129: AssertionError
========================================================================================================= short test summary info ==========================================================================================================
FAILED test_1.py::test_tx_raw_bytes2obj_2 - AssertionError: assert b'' == b'\n\x88\x01\...c\xd3\x8e\x1b'
FAILED test_1.py::test_tx_raw_obj2bytes_2 - AssertionError: assert '0ab0020a8801...010c09a0c1a00' == '0ab0020a8801...010c09a0c0a00'
```


I'll try to create a smaller example, but this is what I actually faced in wild.

**Comment by sterliakov** [comment]
Yeah, actually this reproduce with much shorter example:

```python
from dataclasses import dataclass
from typing import List

import betterproto


@dataclass(eq=False, repr=False)
class TxRaw(betterproto.Message):
    body_bytes: bytes = betterproto.bytes_field(1)
    auth_info_bytes: bytes = betterproto.bytes_field(2)
    signatures: List[bytes] = betterproto.bytes_field(3)


body = bytes([0, 1])
auth = bytes([2, 3])
sig = [b'']


def test_tx_raw_obj2bytes_2():
    # FAILS!
    obj = TxRaw(body_bytes=body, auth_info_bytes=auth, signatures=sig)
    assert TxRaw().parse(bytes(obj)) == obj
```

```
=========================================================================================================== test session starts ============================================================================================================
platform linux -- Python 3.10.2, pytest-7.1.2, pluggy-1.0.0 -- /home/stas/Documents/Work/pyevmos/venv/bin/python
cachedir: .pytest_cache
rootdir: /home/stas/Documents/Work/pyevmos, configfile: pyproject.toml
plugins: web3-5.30.0, cov-3.0.0
collected 19 items / 18 deselected / 1 selected                                                                                                                                                                                            

tests/test_example.py::test_tx_raw_obj2bytes_2 FAILED                                                                                                                                                                                [100%]

================================================================================================================= FAILURES =================================================================================================================
_________________________________________________________________________________________________________ test_tx_raw_obj2bytes_2 __________________________________________________________________________________________________________
tests/test_example.py:123: in test_tx_raw_obj2bytes_2
    assert TxRaw().parse(bytes(obj)) == obj
E   AssertionError: assert TxRaw(body_bytes=b'', auth_info_bytes=b'\x02\x03') == TxRaw(body_bytes=b'\x00\x01', auth_info_bytes=b'\x02\x03', signatures=[b''])
E    +  where TxRaw(body_bytes=b'', auth_info_bytes=b'\x02\x03') = <bound method Message.parse of TxRaw(body_bytes=b'', auth_info_bytes=b'\x02\x03')>(b'\n\x02\x00\x01\x12\x02\x02\x03\n\x00')
E    +    where <bound method Message.parse of TxRaw(body_bytes=b'', auth_info_bytes=b'\x02\x03')> = TxRaw(body_bytes=b'', auth_info_bytes=b'\x02\x03').parse
E    +      where TxRaw(body_bytes=b'', auth_info_bytes=b'\x02\x03') = TxRaw()
E    +    and   b'\n\x02\x00\x01\x12\x02\x02\x03\n\x00' = bytes(TxRaw(body_bytes=b'\x00\x01', auth_info_bytes=b'\x02\x03', signatures=[b'']))
========================================================================================================= short test summary info ==========================================================================================================
FAILED tests/test_example.py::test_tx_raw_obj2bytes_2 - AssertionError: assert TxRaw(body_bytes=b'', auth_info_bytes=b'\x02\x03') == TxRaw(body_bytes=b'\x00\x01', auth_info_bytes=b'\x02\x03', signatures=[b''])
===================================================================================================== 1 failed, 18 deselected in 2.12s =====================================================================================================
```

So `Obj().parse(obj.__bytes__()) != obj`, and this shouldn't happen. This test works for original protobuf:
```python
# -*- coding: utf-8 -*-
# Generated by the protocol buffer compiler.  DO NOT EDIT!
# source: a.proto
"""Generated protocol buffer code."""
from google.protobuf.internal import builder as _builder
from google.protobuf import descriptor as _descriptor
from google.protobuf import descriptor_pool as _descriptor_pool
from google.protobuf import symbol_database as _symbol_database
# @@protoc_insertion_point(imports)

_sym_db = _symbol_database.Default()




DESCRIPTOR = _descriptor_pool.Default().AddSerializedFile(b'\n\x07\x61.proto\"H\n\x05TxRaw\x12\x12\n\nbody_bytes\x18\x01 \x01(\x0c\x12\x17\n\x0f\x61uth_info_bytes\x18\x02 \x01(\x0c\x12\x12\n\nsignatures\x18\x03 \x03(\x0c\x62\x06proto3')

_builder.BuildMessageAndEnumDescriptors(DESCRIPTOR, globals())
_builder.BuildTopDescriptorsAndMessages(DESCRIPTOR, 'a_pb2', globals())
if _descriptor._USE_C_DESCRIPTORS == False:

  DESCRIPTOR._options = None
  _TXRAW._serialized_start=11
  _TXRAW._serialized_end=83
# @@protoc_insertion_point(module_scope)



body = bytes([0, 1])
auth = bytes([2, 3])
sig = [b'']


def test_serialize_and_deserialize():
    # PASSED
    obj = TxRaw(body_bytes=body, auth_info_bytes=auth, signatures=sig)
    assert TxRaw.FromString(obj.SerializeToString()) == obj
```

**Comment by Gobot1234** [comment]
Thank you for your response, the fix you've described seems to work and doesn't break any tests. I'll do a bit of investigation to see if it breaks anything else.

**Comment by sterliakov** [comment]
If that's fine, should I submit a PR with this fix? I think it is a valid solution, because protobuf explicitly allows serializing default/optional fields, even if it is not needed, so the worst thing we may introduce is an increase of the message size. (I mean passing `serialize_empty=True` in `__bool__` method). 

Also, is there any chance it will become new (beta) release in <20 days so that I can use it from PyPi and not from github commit hash? This is needed as a dependency of another PyPi package, so github deps are not allowed.

**Comment by sterliakov** [comment]
BTW, I noticed a place where `betterproto` behaves better than its TypeScript counterpart, giving 2 bytes shorter message (with uint64 field, that can be omitted - confirmed with original protobuf), it's great! Thank you for your efforts:)

**Comment by Gobot1234** [comment]
> If that's fine, should I submit a PR with this fix? 

Be my guest, this should also fix #387

> Also, is there any chance it will become new (beta) release in <20 days

I wouldn't get my hopes up, but I'll see what I can do.
